From e138cecc63034ea39aa1423699f8c0587bc631c8 Mon Sep 17 00:00:00 2001
From: Wang Wensheng <wangwensheng4@huawei.com>
Date: Sat, 8 Apr 2023 01:49:37 +0000
Subject: [PATCH 29/53] [Huawei] mm/sharepool: Change spa_list to rb_tree in
 sp_group

Offering: HULK
hulk inclusion
category: feature
bugzilla: N/A

--------------------------------

When we free/unshare a user address, we first find the corresponding
sp_area from the sp_mapping with the global sp_area_lock. This means
that all the sp_free processes are serial.

However there is a spa_list in each spg and we could search the sp_area
from that list. We should use rb_tree for efficiency.

Actually the spa_list in sp_group is protected by sp_area_lock now and
we would rebuild it later.

Signed-off-by: Wang Wensheng <wangwensheng4@huawei.com>
---
 mm/share_pool.c | 148 +++++++++++++++++++++++++-----------------------
 1 file changed, 77 insertions(+), 71 deletions(-)

diff --git a/mm/share_pool.c b/mm/share_pool.c
index fa54735d74b7..ea8906f91493 100644
--- a/mm/share_pool.c
+++ b/mm/share_pool.c
@@ -162,26 +162,26 @@ struct sp_mapping {
  * the default 8G memory setting for DVPP.
  */
 struct sp_group {
-	int		 id;
-	struct file	 *file;
-	struct file	 *file_hugetlb;
+	int		id;
+	struct file	*file;
+	struct file	*file_hugetlb;
 	/* number of process in this group */
-	int		 proc_num;
+	int		proc_num;
 	/* list head of processes (sp_group_node, each represents a process) */
 	struct list_head proc_head;
-	/* list head of sp_area. it is protected by spin_lock sp_area_lock */
-	struct list_head spa_list;
+	/* it is protected by rw_lock of this spg */
+	struct rb_root	spa_root;
 	/* group statistics */
 	struct sp_meminfo meminfo;
 	/* is_alive == false means it's being destroyed */
-	bool		 is_alive;
-	atomic_t	 use_count;
-	atomic_t	 spa_num;
-	/* protect the group internal elements, except spa_list */
+	bool		is_alive;
+	atomic_t	use_count;
+	atomic_t	spa_num;
+	/* protect the group internal elements */
 	struct rw_semaphore	rw_lock;
 	/* list node for dvpp mapping */
 	struct list_head	mnode;
-	struct sp_mapping       *mapping[SP_MAPPING_END];
+	struct sp_mapping	*mapping[SP_MAPPING_END];
 };
 
 /* a per-process(per mm) struct which manages a sp_group_node list */
@@ -685,7 +685,7 @@ struct sp_area {
 	bool is_dead;
 	atomic_t use_count;		/* How many vmas use this VA region */
 	struct rb_node rb_node;		/* address sorted rbtree */
-	struct list_head link;		/* link to the spg->head */
+	struct rb_node spg_link;	/* link to the spg->rb_root */
 	struct sp_group *spg;
 	enum spa_type type;		/* where spa born from */
 	struct mm_struct *mm;		/* owner of k2u(task) */
@@ -1114,10 +1114,10 @@ static void sp_group_init(struct sp_group *spg, int spg_id)
 	spg->id = spg_id;
 	spg->is_alive = true;
 	spg->proc_num = 0;
+	spg->spa_root = RB_ROOT;
 	atomic_set(&spg->use_count, 1);
 	atomic_set(&spg->spa_num, 0);
 	INIT_LIST_HEAD(&spg->proc_head);
-	INIT_LIST_HEAD(&spg->spa_list);
 	INIT_LIST_HEAD(&spg->mnode);
 	init_rwsem(&spg->rw_lock);
 	meminfo_init(&spg->meminfo);
@@ -1327,6 +1327,7 @@ static struct sp_group_node *group_add_task(struct mm_struct *mm, struct sp_grou
  */
 int mg_sp_group_add_task(int tgid, unsigned long prot, int spg_id)
 {
+	struct rb_node *p;
 	struct task_struct *tsk;
 	struct mm_struct *mm;
 	struct sp_group *spg;
@@ -1460,8 +1461,8 @@ int mg_sp_group_add_task(int tgid, unsigned long prot, int spg_id)
 	 * new process' page table.
 	 */
 	spin_lock(&sp_area_lock);
-
-	list_for_each_entry(spa, &spg->spa_list, link) {
+	for (p = rb_first(&spg->spa_root); p; p = rb_next(p)) {
+		spa = container_of(p, struct sp_area, spg_link);
 		__sp_area_drop_locked(prev);
 		prev = spa;
 
@@ -1545,27 +1546,34 @@ int mg_sp_id_of_current(void)
 }
 EXPORT_SYMBOL_GPL(mg_sp_id_of_current);
 
+#define insert_sp_area(__root, __spa, node, rb_node_param)	\
+	do {							\
+		struct rb_node **p = &((__root)->rb_node);	\
+		struct rb_node *parent = NULL;			\
+		while (*p) {					\
+			struct sp_area *tmp;			\
+			parent = *p;				\
+			tmp = rb_entry(parent, struct sp_area, rb_node_param);\
+			if (__spa->va_start < tmp->va_end)	\
+				p = &(*p)->rb_left;		\
+			else if (__spa->va_end > tmp->va_start)	\
+				p = &(*p)->rb_right;		\
+			else					\
+				BUG();				\
+		}						\
+		rb_link_node((node), parent, p);		\
+		rb_insert_color((node), (__root));		\
+	} while (0)
+
 /* the caller must hold sp_area_lock */
-static void insert_sp_area(struct sp_mapping *spm, struct sp_area *spa)
+static void spm_insert_area(struct sp_mapping *spm, struct sp_area *spa)
 {
-	struct rb_node **p = &spm->area_root.rb_node;
-	struct rb_node *parent = NULL;
-
-	while (*p) {
-		struct sp_area *tmp;
-
-		parent = *p;
-		tmp = rb_entry(parent, struct sp_area, rb_node);
-		if (spa->va_start < tmp->va_end)
-			p = &(*p)->rb_left;
-		else if (spa->va_end > tmp->va_start)
-			p = &(*p)->rb_right;
-		else
-			BUG();
-	}
+	insert_sp_area(&spm->area_root, spa, &spa->rb_node, rb_node);
+}
 
-	rb_link_node(&spa->rb_node, parent, p);
-	rb_insert_color(&spa->rb_node, &spm->area_root);
+static void sp_group_insert_area(struct sp_group *spg, struct sp_area *spa)
+{
+	insert_sp_area(&spg->spa_root, spa, &spa->spg_link, spg_link);
 }
 
 /**
@@ -1714,11 +1722,11 @@ static struct sp_area *sp_alloc_area(unsigned long size, unsigned long flags,
 	spa->preferred_node_id = node_id;
 	spa->device_id = device_id;
 
-	insert_sp_area(mapping, spa);
+	spm_insert_area(mapping, spa);
 	mapping->free_area_cache = &spa->rb_node;
-	list_add_tail(&spa->link, &spg->spa_list);
 
 	spin_unlock(&sp_area_lock);
+	sp_group_insert_area(spg, spa);
 
 	spa_inc_usage(spa);
 	return spa;
@@ -1729,40 +1737,34 @@ static struct sp_area *sp_alloc_area(unsigned long size, unsigned long flags,
 	return err;
 }
 
-/* the caller should hold sp_area_lock */
-static struct sp_area *find_sp_area_locked(struct sp_group *spg,
+/*
+ * Find a spa with key @addr from @spg and increase its use_count.
+ * The caller should hold spg->rw_lock
+ */
+static struct sp_area *sp_area_get(struct sp_group *spg,
 		unsigned long addr)
 {
-	struct sp_mapping *spm = sp_mapping_find(spg, addr);
-	struct rb_node *n = spm->area_root.rb_node;
+	struct rb_node *n = spg->spa_root.rb_node;
 	while (n) {
 		struct sp_area *spa;
 
-		spa = rb_entry(n, struct sp_area, rb_node);
+		spa = rb_entry(n, struct sp_area, spg_link);
 		if (addr < spa->va_start) {
 			n = n->rb_left;
 		} else if (addr > spa->va_start) {
 			n = n->rb_right;
 		} else {
-			return spa;
+			/* a spa without any user will die soon */
+			if (atomic_inc_not_zero(&spa->use_count))
+				return spa;
+			else
+				return NULL;
 		}
 	}
 
 	return NULL;
 }
 
-static struct sp_area *get_sp_area(struct sp_group *spg, unsigned long addr)
-{
-	struct sp_area *n;
-
-	spin_lock(&sp_area_lock);
-	n = find_sp_area_locked(spg, addr);
-	if (n)
-		atomic_inc(&n->use_count);
-	spin_unlock(&sp_area_lock);
-	return n;
-}
-
 static bool vmalloc_area_clr_flag(unsigned long kva, unsigned long flags)
 {
 	struct vm_struct *area;
@@ -1813,7 +1815,6 @@ static void sp_free_area(struct sp_area *spa)
 		pr_debug("clear spa->kva %ld is not valid\n", spa->kva);
 
 	spa_dec_usage(spa);
-	list_del(&spa->link);
 
 	rb_erase(&spa->rb_node, &spm->area_root);
 	RB_CLEAR_NODE(&spa->rb_node);
@@ -1831,8 +1832,12 @@ static void __sp_area_drop_locked(struct sp_area *spa)
 	if (!spa)
 		return;
 
-	if (atomic_dec_and_test(&spa->use_count))
+	if (atomic_dec_and_test(&spa->use_count)) {
+		down_write(&spa->spg->rw_lock);
+		rb_erase(&spa->spg_link, &spa->spg->spa_root);
+		up_write(&spa->spg->rw_lock);
 		sp_free_area(spa);
+	}
 }
 
 static void __sp_area_drop(struct sp_area *spa)
@@ -1939,8 +1944,10 @@ static int sp_free_inner(unsigned long addr, int spg_id, bool is_sp_free)
 		return -EINVAL;
 	}
 
-	spa = get_sp_area(spg, addr);
+	down_write(&spg->rw_lock);
+	spa = sp_area_get(spg, addr);
 	if (!spa) {
+		up_write(&spg->rw_lock);
 		pr_debug("%s, invalid input addr %lx\n", str, addr);
 		ret = -EINVAL;
 		goto out;
@@ -1948,12 +1955,12 @@ static int sp_free_inner(unsigned long addr, int spg_id, bool is_sp_free)
 
 	if ((is_sp_free && spa->type != SPA_TYPE_ALLOC) ||
 	    (!is_sp_free && spa->type == SPA_TYPE_ALLOC)) {
+		up_write(&spg->rw_lock);
 		ret = -EINVAL;
 		pr_warn("%s failed, spa_type is not correct\n", str);
 		goto drop_spa;
 	}
 
-	down_write(&spa->spg->rw_lock);
 	if (!current->mm && spa->applier != current->tgid) {
 		up_write(&spg->rw_lock);
 		ret = -EPERM;
@@ -1962,8 +1969,8 @@ static int sp_free_inner(unsigned long addr, int spg_id, bool is_sp_free)
 		goto drop_spa;
 	}
 
-	if (!spg_valid(spa->spg)) {
-		up_write(&spa->spg->rw_lock);
+	if (!spg_valid(spg)) {
+		up_write(&spg->rw_lock);
 		if (spa->type != SPA_TYPE_ALLOC)
 			vmalloc_area_clr_flag(spa->kva, VM_SHAREPOOL);
 		goto drop_spa;
@@ -1972,16 +1979,16 @@ static int sp_free_inner(unsigned long addr, int spg_id, bool is_sp_free)
 
 	/* the life cycle of spa has a direct relation with sp group */
 	if (unlikely(spa->is_dead)) {
-		up_write(&spa->spg->rw_lock);
+		up_write(&spg->rw_lock);
 		pr_err_ratelimited("%s, unexpected double sp free\n", str);
 		dump_stack();
 		ret = -EINVAL;
 		goto drop_spa;
 	}
 	spa->is_dead = true;
-	up_write(&spa->spg->rw_lock);
+	up_write(&spg->rw_lock);
 
-	down_read(&spa->spg->rw_lock);
+	down_read(&spg->rw_lock);
 	__sp_free(spa, NULL);
 	if (spa->type == SPA_TYPE_ALLOC) {
 		sp_fallocate(spa);
@@ -1989,7 +1996,7 @@ static int sp_free_inner(unsigned long addr, int spg_id, bool is_sp_free)
 		vmalloc_area_clr_flag(spa->kva, VM_SHAREPOOL);
 		spa->kva = 0;
 	}
-	up_read(&spa->spg->rw_lock);
+	up_read(&spg->rw_lock);
 
 	if (current->mm != NULL)
 		sp_update_process_stat(current, false, spa);
@@ -2308,9 +2315,9 @@ void *__mg_sp_alloc_nodemask(unsigned long size, unsigned long sp_flags, int spg
 		return ERR_PTR(-ENODEV);
 	}
 
-	down_read(&spg->rw_lock);
+	down_write(&spg->rw_lock);
 	if (!spg_valid(spg)) {
-		up_read(&spg->rw_lock);
+		up_write(&spg->rw_lock);
 		sp_group_put(spg);
 		pr_info("allocation failed, spg is dead\n");
 		return ERR_PTR(-EINVAL);
@@ -2319,7 +2326,7 @@ void *__mg_sp_alloc_nodemask(unsigned long size, unsigned long sp_flags, int spg
 	spa = sp_alloc_area(ac.size_aligned, ac.sp_flags, spg,
 			    ac.type, current->tgid, ac.preferred_node_id);
 	if (IS_ERR(spa)) {
-		up_read(&spg->rw_lock);
+		up_write(&spg->rw_lock);
 		pr_err("alloc spa failed in allocation(potential no enough virtual memory when -75): %ld\n",
 			PTR_ERR(spa));
 		ret = PTR_ERR(spa);
@@ -2330,8 +2337,7 @@ void *__mg_sp_alloc_nodemask(unsigned long size, unsigned long sp_flags, int spg
 	if (ret == -ENOMEM && sp_alloc_fallback(spa, &ac))
 		goto try_again;
 
-	up_read(&spg->rw_lock);
-
+	up_write(&spg->rw_lock);
 	if (!ret)
 		sp_update_process_stat(current, true, spa);
 
@@ -2561,10 +2567,10 @@ void *mg_sp_make_share_k2u(unsigned long kva, unsigned long size,
 		return ERR_PTR(-ENODEV);
 	}
 
-	down_read(&spg->rw_lock);
+	down_write(&spg->rw_lock);
 	spa = sp_alloc_area(kc.size_aligned, kc.sp_flags, spg, kc.type, current->tgid, 0);
 	if (IS_ERR(spa)) {
-		up_read(&spg->rw_lock);
+		up_write(&spg->rw_lock);
 		pr_err("alloc spa failed in k2u_spg (potential no enough virtual memory when -75): %ld\n",
 				PTR_ERR(spa));
 		sp_group_put(spg);
@@ -2587,7 +2593,7 @@ void *mg_sp_make_share_k2u(unsigned long kva, unsigned long size,
 		ret = mmap_ret;
 	}
 
-	up_read(&spg->rw_lock);
+	up_write(&spg->rw_lock);
 	if (!ret) {
 		sp_update_process_stat(current, true, spa);
 		vmalloc_area_set_flag(kc.kva_aligned, VM_SHAREPOOL);
-- 
2.17.1

